design/history/exploration-reports/2021.02-cross-language-fixtures.md

```diff
@@ -0,0 +1,160 @@
+	- if this is supported, ideally we'll have a check in CI that makes sure all files are canonically formatted.
+- stretch goal: handles content-addressable documents well.
+	- this is a stretch goal rather than a hard requirement because: content-addressable DAGs are _common_ in our work, but not _universal_.  Some things we write fixtures for don't involve DAGs, so we should solve that problem first, and then figure out how to also handle DAGs.
+	- this is a stretch goal rather than a hard requirement because: this author has the hypothesis this will be easy to add to existing container formats.
+
+
+prior art
+---------
+
+- selector fixtures in json: https://github.com/ChainSafe/ipld-traversal-vectors
+	- notes on outcome: awesome... except when people eyeball it, the fact that the top level maps are containing "control plane" info (which part of the fixture it is) is non-obvious.
+- schema DSL parser fixtures: https://github.com/ipld/js-ipld-schema/tree/master/test/fixtures and https://github.com/ipld/go-ipld-schema/tree/master/test/fixtures
+	- notes on outcome: the parts that use yaml: awesome... although a lot of the yaml files are resorting to heredocs.
+		- heredocs aren't always preserved by yaml parse and reserialize, so that's a knock against the "programatically updatable" goal.
+		- This author is frankly just not a fan of yaml.  The parsers are... complex.
+	- notes on outcome: the parts that don't use yaml: awesome... except sometimes there's two files that are siblings, and they're related, and you just have to "know" that.
+- (other project) "interactive" fixtures in noise: https://github.com/pipedown/noise/blob/master/repl-tests/limit.noise
+	- it has a fixture update script!  (yes!)  https://github.com/pipedown/noise/blob/master/update-test-repl.sh
+	- interesting example.  leans a lot on domain specifics for separation, but interesting.
+	- hypothetically, the fact that these are commands to a textual interactive UI means it could also have included other binary files by commands in the system which do so.  (doesn't generalize to us, but interesting.)
+- (other project) fixtures in repeatr: https://github.com/polydawn/repeatr/blob/master/examples/hello-cached.tcase
+	- this is using wishfix, which will be discussed more below
+	- as with the above example, there are auto-updating scripts, so you write some of the fixture, and then run it, and use git to control changes in the future.
+	- as with the above example, this happened to "run commands", and it does actually reference other files.  (probably doesn't generalize to IPLD needs, but interesting.)
+
+
+options
+-------
+
+### txtar
+
+Txtar is a simple textual format, with reference implementation at https://pkg.go.dev/golang.org/x/tools/txtar .
+
+- Pros:
+	- fits the requirements.
+	- needs verification: clarity of trailing linebreaks.
+- Cons:
```

Comment:
It's probably fine if you're careful since it's obvious, but from what I can tell your fixture cannot contain the characters `-- test --` since it will be treated as if you're starting a new fixture and there are no escape sequences defined in the format.

Reply:
Glad you pointed this out.  I was so enthusiastic about txtar on my first glance at it that I somehow forgot to check this.

Updated document with notes about band size for all the options.